### PR TITLE
feat: implement isolated hooks

### PR DESCRIPTION
**Pull request by isidentical**
This PR implements "isolated hooks" (e.g., hooks which run in a different environment than the fal itself). Current implementation:

- Only uses virtual envs, so only requirements that can be installed with pip is available
- Virtual envs that are created for an environment set is cached. So in the second run for the same hook (or a different hook that has the same exact dependencies) it will be very fast.
- Each environment contains `fal` + `dbt-core` and all the dbt adapters you have locally (we scan all the installed python packages you have, and install the same version of every `dbt-*` adapter)

### Example
```yaml
  - name: model_b
    meta:
      fal:
        pre-hook:
         # This is a regular hook, that runs on the same environment as the fal itself
          - path: local_hook.py

          # This is an isolated hook, which runs in an environment called 'funny'. These
          # environments are defined inside fal_project.yaml, and uses the same format
          # as the "fal runtime" (so they are interchangeable).
          - path: funny_hook.py
            environment: funny
           
          # You can specify arguments just like you could for local hooks, and since
          # arguments are yaml values the serialization is very easy!
          - path: check_imports.py
            environment: not-funny
            with:
              import: pyjokes
              expected_success: false
```

Here is the environment spec:
```yaml
environments:
  # An environment without a 'requirements' key is considered a bare virtual
  # environment where you want to run your code without the affect of other
  # threads. This one only includes fal + dbt-core + dbt-* adapters
  - name: not-funny
  
  # This one defines some requirements, so they are also installed in addition to the
  # defined requirements above.
  - name: funny
    requirements:
      - pyjokes==0.6.0  
```

### CODE REVIEW

**Comment by isidentical** [comment]
Requires #487 and #491

File: integration_tests/projects/013_structured_hooks/models/schema.yml
Comment (mederka) [inline integration_tests/projects/013_structured_hooks/models/schema.yml:44]: atl: what's the benefit of specifying imports here?

**Review by mederka — APPROVED** [review]
This is great!

File: src/fal/packages/environment.py
Comment (chamini2) [inline src/fal/packages/environment.py:112]: ATL: why are we printing instead of using a logger? Is it not available in this context?

File: src/fal/packages/environment.py
Comment (chamini2) [inline src/fal/packages/environment.py:150]: ATL: why do we do a subprocess instead of running in this same process? it seems we are already in a thread, right?

File: src/faldbt/project.py
Comment (chamini2) [inline src/faldbt/project.py:444]: this is surprisingly small

**Review by chamini2 — COMMENTED** [review]
Please don't merge until we discuss these. Thanks for very interesting PR!

File: src/fal/packages/_run_hook.py
Comment (chamini2) [reply src/fal/packages/_run_hook.py:17]: https://github.com/fal-ai/fal/blob/c70a2f4b37621accbd256d85de5e5181e866cf16/src/fal/cli/cli.py#L33-L38

File: src/faldbt/project.py
Comment (isidentical) [reply src/faldbt/project.py:444]: Not sure if there is any missing one, but these are all I found.

File: integration_tests/projects/013_structured_hooks/models/schema.yml
Comment (isidentical) [reply integration_tests/projects/013_structured_hooks/models/schema.yml:44]: Just to play with the combinations of arguments to hooks, and writing a hook that can actually be a test of itself.

File: src/fal/packages/_run_hook.py
Comment (isidentical) [reply src/fal/packages/_run_hook.py:1]: nop!

File: src/fal/packages/environment.py
Comment (chamini2) [reply src/fal/packages/environment.py:91]: OK, I would say let's ignore fal or dbt versions and log a warning for them

File: src/fal/packages/environment.py
Comment (chamini2) [reply src/fal/packages/environment.py:54]: > I wish it was part of the public API though, we could have just used it instead of copying this complex snippet.

I know!

File: src/fal/packages/environment.py
Comment (chamini2) [reply src/fal/packages/environment.py:154]: Of course! This makes a lot of sense. Very cool experiments, though.
I agree that it is very complex without really adding substantial value. Not worth doing.

Thanks for the explanation!

File: src/fal/packages/_run_hook.py
Comment (chamini2) [reply src/fal/packages/_run_hook.py]: > ATIL: what is new dbt logging?

https://github.com/dbt-labs/dbt-core/blob/v1.0.0/core/dbt/events/README.md#adapter-maintainers

We had not migrated because 0.X dbt did not have the new logging available. I will get it done soon!

Comment (chamini2) [reply src/fal/packages/_run_hook.py]: I think the initial idea of just adding a `TODO` comment in here to remember to take a look is enough. To not block this PR at all.

File: src/fal/packages/environment.py
Comment (chamini2) [inline src/fal/packages/environment.py:99]: should we run this only once in the `__post_init__` and store the result? to mimic the past `cached_property`?

Comment (chamini2) [inline src/fal/packages/environment.py:99]: should we run this only once in the `__post_init__` and store the result? to mimic the past `cached_property`?

Comment (isidentical) [reply src/fal/packages/environment.py:99]: I've reduced the number of `_key` accesses to one so it is probably not necessary.

File: src/fal/packages/_run_hook.py
Comment (isidentical) [reply src/fal/packages/_run_hook.py]: I've unified both fal.cli and this part, so once this is landed and the changes are applied it will fix both of them I think. Thanks for mentioning about this btw!

File: src/fal/packages/environment.py
Comment (isidentical) [reply src/fal/packages/environment.py:92]: The parsing of dependencies would be a bit tricky (`fal==5.0`, `fal!=5.0`, and a lot of other variations) though we can just let pip handle it. It already raises a warning if you have a dependency conflict, and I think it will handle this case too. If it starts creating confusions, we can think about it later I think.
 
```console
 $ pip install pyjokes==0.5.0 pyjokes==0.6.0                                                                                                                                                                                                                                         5h54m
Collecting pyjokes==0.5.0
  Using cached pyjokes-0.5.0-py2.py3-none-any.whl (12 kB)
ERROR: Cannot install pyjokes==0.5.0 and pyjokes==0.6.0 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested pyjokes==0.5.0
    The user requested pyjokes==0.6.0
```

Comment (chamini2) [reply src/fal/packages/environment.py:92]: Cool, let's do that!

File: src/fal/packages/environment.py
Comment (isidentical) [reply src/fal/packages/environment.py]: Done! (and used a different method for the rest [fal, dbt-core] so we can drop the dependency on importlib-metadata).

File: src/fal/packages/environment.py
Comment (isidentical) [reply src/fal/packages/environment.py]: @chamini2 I just tried and noticed that since the dataclass if frozen, we can't do it on the `__post_init__` 😄 As I said, it is extremely fast and side-effect free, so let's not spend too much time on it.

**Comment by isidentical** [comment]
Thanks a lot for the reviews guys! @chamini2 please re-review, all the reviews should be addressed now.

File: src/fal/packages/environment.py
Comment (chamini2) [inline src/fal/packages/environment.py:82]: Probably needs to change before merge?

File: src/fal/packages/environment.py
Comment (chamini2) [inline src/fal/packages/environment.py:46]: Add a link to the file in the repo
```suggestion
# NOTE: This is from dbt https://github.com/dbt-labs/dbt-core/blob/7bd861a3514f70b64d5a6c642b4204b50d0d3f7e/core/dbt/version.py#L209-L236
```

**Review by chamini2 — APPROVED** [review]
Awesome, I think that is it!

File: src/fal/packages/environment.py
Comment (isidentical) [reply src/fal/packages/environment.py:82]: Yep, awaiting the proper versions PR

Comment (isidentical) [reply src/fal/packages/environment.py]: Fixed!

**Pull request merged**
